pkg/kubelet/kubelet.go

```diff
@@ -557,11 +558,78 @@ func (kl *Kubelet) Run(updates <-chan PodUpdate) {
 	kl.syncLoop(updates, kl)
 }
 
+func (kl *Kubelet) initialNodeStatus() (*api.Node, error) {
+	node := &api.Node{
+		ObjectMeta: api.ObjectMeta{Name: kl.hostname},
+		Status: api.NodeStatus{
+			Capacity: api.ResourceList{
+				api.ResourceCPU:    *resource.NewMilliQuantity(1000, resource.DecimalSI),
+				api.ResourceMemory: resource.MustParse("3Gi"),
+			},
+		},
+	}
+	if kl.cloud != nil {
+		instances, ok := kl.cloud.Instances()
+		if !ok {
+			return nil, fmt.Errorf("failed to get instances from cloud provider")
+		}
+		// TODO(roberthbailey): Inspect the local machine rather than calling out
+		// to the cloud provider.
+		resources, err := instances.GetNodeResources(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node resources from cloud provider: %v", err)
+		}
+		if resources != nil {
+			node.Status.Capacity = resources.Capacity
+		}
+		instanceID, err := instances.ExternalID(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get instance ID from cloud provider: %v", err)
+		}
+		node.Spec.ExternalID = instanceID
+		nodeAddresses, err := instances.NodeAddresses(kl.hostname)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node address from cloud provider: %v", err)
+		}
+		node.Status.Addresses = nodeAddresses
+	}
+	return node, nil
+}
+
+// registerNode registers the node with the cluster master.
+func (kl *Kubelet) registerNode() error {
```

Comment:
This change will make clustering much, much simpler. I haven't looked at kube-register in detail, but this may make it unnecessary because as long as you point kubelets at the master they will join the cluster and become schedulable.